pkg/scanner/validation_test.go

```diff
@@ -0,0 +1,543 @@
+			expression: "namespaces.items.all(ns, pods.items.exists(pod, pod.metadata.namespace == ns.metadata.name))",
+			declarations: []*expr.Decl{
+				decls.NewVar("namespaces", decls.Dyn),
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: true,
+		},
+		{
+			name:       "undeclared variable in expression",
+			expression: "deployments.items.all(d, d.spec.replicas > 1)",
+			declarations: []*expr.Decl{
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: false,
+			wantError: ValidationErrorTypeUndeclaredReference,
+		},
+		{
+			name:       "valid CEL macros - all",
+			expression: "pods.items.all(pod, pod.spec.containers.all(c, c.image != ''))",
+			declarations: []*expr.Decl{
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: true,
+		},
+		{
+			name:       "valid CEL macros - exists",
+			expression: "pods.items.exists(pod, pod.metadata.labels.app == 'test')",
+			declarations: []*expr.Decl{
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: true,
+		},
+		{
+			name:       "valid CEL macros - filter",
+			expression: "pods.items.filter(pod, pod.status.phase == 'Running').size() > 0",
+			declarations: []*expr.Decl{
+				decls.NewVar("pods", decls.Dyn),
+			},
+			wantValid: true,
+		},
+	}
+
+	validator := NewRuleValidator(nil)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			issues := validator.ValidateCELExpression(tt.expression, tt.declarations)
+
+			if tt.wantValid {
+				if len(issues) > 0 {
+					t.Errorf("Expected valid expression, got issues: %v", issues)
+				}
+			} else {
+				if len(issues) == 0 {
+					t.Errorf("Expected validation errors, but got none")
+				} else if tt.wantError != "" && issues[0].Type != tt.wantError {
+					t.Errorf("Expected error type %s, got %s", tt.wantError, issues[0].Type)
+				}
+			}
+		})
+	}
+}
+
+// Test CompileCELExpression public function
+func TestCompileCELExpression(t *testing.T) {
+	tests := []struct {
+		name       string
+		expression string
+		inputs     []Input
+		wantError  bool
+		errorMsg   string
+	}{
+		{
+			name:       "valid expression with inputs",
+			expression: "deployments.items.all(d, d.spec.replicas >= 2)",
+			inputs: []Input{
+				&InputImpl{
+					InputName: "deployments",
+					InputType: InputTypeKubernetes,
+				},
+			},
+			wantError: false,
+		},
+		{
+			name:       "expression with undeclared variable",
+			expression: "pods.items.all(p, p.spec.replicas > 0)",
+			inputs: []Input{
+				&InputImpl{
+					InputName: "deployments",
+					InputType: InputTypeKubernetes,
+				},
+			},
+			wantError: true,
+			errorMsg:  "UNDECLARED_REFERENCE",
+		},
+		{
+			name:       "syntax error in expression",
+			expression: "pods.items.all(p, p.spec.replicas >",
+			inputs: []Input{
+				&InputImpl{
+					InputName: "pods",
+					InputType: InputTypeKubernetes,
+				},
+			},
+			wantError: true,
+			errorMsg:  "SYNTAX_ERROR",
+		},
+		{
+			name: "complex valid expression",
```

Comment:
In particular, this is complex because it's using multiple membership checks (`has()`) inside a containment check (`all())`). We can elaborate on that in the test name to make it more specific - since `complex` can mean different things to different readers.